draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -2712,28 +2710,131 @@ of the counter byte I2OSP(i, 1)).
 
 The essential difference between the construction of {{CDMP05}} and
 expand\_message\_xmd is that the latter hashes a counter appended to
-strxor(b\_0, b\_(i - 1)) (step 9) rather than to b\_0.
+strxor(b\_0, b\_(i - 1)) (step 10) rather than to b\_0.
 This approach increases the Hamming distance between inputs to different
 invocations of H, which reduces the likelihood that nonidealities in H
 affect the distribution of the b\_i values.
 
+## Domain separation recommendations {#security-considerations-domain-separation}
+
+The expand\_message variants in this document ({{hashtofield-expand}})
+always append a suffix-free-encoded domain separation tag DST\_prime
+to the strings hashed by H, the underlying hash or extensible output function.
+(Other expand\_message variants that follow the guidelines in
+{{hashtofield-expand-other}} are expected to have similar properties.
+However, these should be analyzed on a case-by-case basis.)
+
+For protocols that also use H outside of hash\_to\_field, those uses
+MUST be domain separated from all uses of H inside hash\_to\_field, and
+SHOULD be domain separated from uses of H outside of the protocol.
+The RECOMMENDED method of ensuring domain separation is appending a tag
+distinct from DST\_prime to all inputs to H outside of hash\_to\_field.
+This ensures that distinct uses of H have distinct suffixes.
+If further separation is required among uses of H outside of
+hash\_to\_field, protocols should choose multiple tags that are
+all distinct from one another and from DST\_prime.
+
+The above method is not compatible with all uses of H.
+For example, this method does not work with HMAC {{RFC2104}},
+because there is no way to append a value to the input of
+the outer call to H.
+The following alternative methods MAY be used instead of the
+recommended method given above.
+We discuss the specific case of HMAC further below.
+
+1. For each use of H outside hash\_to\_field, choose a unique domain
+   separation tag DST\_ext. Augment each invocation of H on input msg by
+   computing H(DST\_ext || msg || I2OSP(0, 1)).
+
+   This ensures domain separation as follows.
+   First, appending the byte I2OSP(0, 1) to all inputs to H outside hash\_to\_field
+   ensures that these inputs are distinct from those inside hash\_to\_field
+   because the final byte of DST\_prime encodes the length of DST, which is
+   required to be nonzero ({{domain-separation}}, requirement 2), and DST\_prime
+   is always appended to invocations of H inside hash\_to\_field.
+   Second, distinct DST\_ext values ensure that uses of H outside
+   hash\_to\_field are domain separated from one another.
+
+   For example, for two uses of H whose inputs are msg1 and msg2,
```

Comment:
I think you might want to say two hash functions H0, H1 created from H. So each hash function has different `DST_ext` prefix.

Otherwise, this makes me think that every time Alice wants to hash using say H0 (derived from H), she needs to append a different DST_ext (per message). Rather than a DST_ext that is the same for all invocations of H0.

Reply:
Wouldn't those be identical? Namely, Alice could append the DST, or the implementation of H0 could do it internally, no?

Reply:
The difference is whether the `DTS_ext` is different per message rather than per hash H0,H1,...

Reply:
Sorry, I'm not following. How are these not identical? 

```
H0(msg) = H(msg || DST1) // Library appends and calls H directly
H1(msg) = H(msg || DST2) // ^^
```

and

```
H0(msg || DST1) = H(msg || DST1) // Alice appends and calls H directly
H1(msg || DST2) = H(msg || DST2) // ^^
```

Reply:
This is not the point. 

> For example, for two uses of H whose inputs are msg1 and msg2, one might choose distinct DST\_ext1 and DST\_ext2 and compute
>
>        hash1 = H(DST_ext1 || msg1 || I2OSP(0, 1))
>        hash2 = H(DST_ext2 || msg2 || I2OSP(0, 1))

Every time Alice wants to hash with H0 defined as H0(msg) = H( DST_ex || msg ),  she must choose a different DST_ex. 

So, which one is the correct way?
H0(msg1) = H( DST_ex1 || msg1 )
H0(msg2) = H( DST_ex2 || msg2 )
H0(msg3) = H( DST_ex3 || msg3 )
or this one
H0(msg1) = H( DST_ex || msg1 )
H0(msg2) = H( DST_ex || msg2 )
H0(msg3) = H( DST_ex || msg3 )

I think the correct way is the second one, but the text it seems to me is saying the first one.

Reply:
Ah, I see! Certainly the second one is the recommendation here. Maybe this clears it up?

```
hash1 = H1(msg) = H(DST_ext1 || msg || I2OSP(0, 1))
hash2 = H2(msg) = H(DST_ext2 || msg || I2OSP(0, 1))
```

Reply:
@kwantam if this works for you, would you mind adding it? I'll merge when done!

Reply:
Great catch. I will clarify.